stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -198,31 +198,71 @@ Base.literal_pow(::typeof(^), D::Diagonal, valp::Val) =
     Diagonal(Base.literal_pow.(^, D.diag, valp)) # for speed
 Base.literal_pow(::typeof(^), D::Diagonal, ::Val{-1}) = inv(D) # for disambiguation
 
+function _muldiag_size_check(A, B)
+    nA = size(A, 2)
+    mB = size(B, 1)
+    @noinline throw_dimerr(::AbstractMatrix, nA, mB) = throw(DimensionMismatch("second dimension of A, $nA, does not match first dimension of B, $mB"))
+    @noinline throw_dimerr(::AbstractVector, nA, mB) = throw(DimensionMismatch("second dimension of D, $nA, does not match length of V, $mB"))
+    nA == mB || throw_dimerr(B, nA, mB)
+    return nothing
+end
+function _muldiag_size_check(C, A, B)
+    _muldiag_size_check(A, B)
+    @noinline throw_dimerr(::AbstractMatrix, szC, szA) = throw(DimensionMismatch("output matrix has size: $szC, but should have size $szA"))
+    @noinline throw_dimerr(::AbstractVector, szC, szA) = throw(DimensionMismatch("output matrix has size: $szC, but should have size $szA"))
+    _size_check_out(C, ::Diagonal, A) = _size_check_out(C, A)
+    _size_check_out(C, A, ::Diagonal) = _size_check_out(C, A)
+    _size_check_out(C, A::Diagonal, ::Diagonal) = _size_check_out(C, A)
+    function _size_check_out(C, A)
+        szA = size(A)
+        szC = size(C)
+        szA == szC || throw_dimerr(B, szC, szA)
+        return nothing
+    end
+    _size_check_out(C, A, B)
+end
+
 function (*)(Da::Diagonal, Db::Diagonal)
-    nDa, mDb = size(Da, 2), size(Db, 1)
-    if nDa != mDb
-        throw(DimensionMismatch("second dimension of Da, $nDa, does not match first dimension of Db, $mDb"))
-    end
+    _muldiag_size_check(Da, Db)
     return Diagonal(Da.diag .* Db.diag)
 end
 
 function (*)(D::Diagonal, V::AbstractVector)
-    nD = size(D, 2)
-    if nD != length(V)
-        throw(DimensionMismatch("second dimension of D, $nD, does not match length of V, $(length(V))"))
-    end
+    _muldiag_size_check(D, V)
     return D.diag .* V
 end
 
+# Optimized multiplication between Triangular and Diagonal matrices
```

Comment:
An extended version of this, which includes `/` related calls, is under going in #42343. So these 2 PRs seem to conflict with each other.

Reply:
Ah I see, so maybe I'll remove these from this PR

Reply:
I've removed these from this PR